documentation: add release documentation (#1458)
quay/claircore: docs/contributor/releases.md

# Releases

> Claircore releases are cut as and when they are needed.

Reviewer: ```suggestion
Claircore releases are cut when they are needed, as judged by the maintainers.
```

> Traditionally releases are always cut from the `main` branch so as not to complicate
> the release process, however, on rare occasions when a fix is time-sensitive then
> cutting a release from a release branch is possible.

Reviewer: ```suggestion
Releases are made from the `main` branch.
On rare occasions when a fix is time-sensitive, it is possible to create a release branch and make a release from it.
```

## Process

> **_NOTE:_** Ensure changelog entries have been created for the relevant commits.
> (see [Changelog documentation](./changelog.md))

### From main

> ```sh
> .github/scripts/prepare-release -b main -r upstream v1.x.x
> ```

Reviewer: ```suggestion
```sh
NEW_VERSION=v0.999.999
.github/scripts/prepare-release -b main -r upstream "$NEW_VERSION"
```
```

Follow the `prepare-release` command's instructions to merge changelog updates and release the tag.

### From release branch

First create the relevant release branch (e.g if you are releasing `v1.6.1` create `release-1.6` from the
previous tag (in this case `v1.6.0`)). Then backport any commits and push up the release branch.

> ```sh
> git branch release-1.x v1.x.x
> git cherry-pick -x {commit sha}
> git push upstream release-1.x
> ```

Reviewer: ```suggestion
```sh
LAST_MINOR=v0.999.0
BRANCH=release-${LAST_MINOR%.*} # e.g. release-v0.999
git branch $BRANCH $LAST_MINOR
TO_BACKPORT=beefc0ffee # Use the commit digest of the original commit
git cherry-pick -x $TO_BACKPORT
git push upstream $BRANCH
```
```

Finally, prepare the release specifying the release branch.

> ```sh
> .github/scripts/prepare-release -b release-1.x -r upstream v1.x.x
> ```

Reviewer: ```suggestion
```sh
LAST_MINOR=v0.999.0
NEW_VERSION=v0.999.1
BRANCH=release-${LAST_MINOR%.*} # e.g. release-v0.999
.github/scripts/prepare-release -b $BRANCH -r upstream $NEW_VERSION
```
```

# Releases

Claircore releases are cut when they are needed, as judged by the maintainers.

Releases are made from the `main` branch.
On rare occasions when a fix is time-sensitive, it is possible to create a release branch and make a release from it.

## Process

> **_NOTE:_** Ensure changelog entries have been created for the relevant commits.
> (see [Changelog documentation](./changelog.md))

### From main

```sh
NEW_VERSION=v0.999.999
.github/scripts/prepare-release -b main -r upstream "$NEW_VERSION"
```

Follow the `prepare-release` command's instructions to merge changelog updates and release the tag.

### From release branch

> First create the relevant release branch (e.g if you are releasing `v1.6.1` create `release-1.6` from the
> previous tag (in this case `v1.6.0`)). Then backport any commits and push up the release branch.

Reviewer: ```suggestion
First, create the relevant release branch.
For example, if you are releasing `v1.6.1` create `release-1.6` from the
previous tag (in this case, `v1.6.0`).

Next, cherry-pick any needed commits with the `-x` flag to keep a reference to the original commit.
This may involve rewriting the changes.

Once the backports are done, push the release branch.
```

Author: I'm going to update this to reflect the version/branch examples from the snippets
